x/wasm/IBC.md

```diff
@@ -0,0 +1,128 @@
+# IBC specification
+
+This documents how CosmWasm contracts are expected to interact with IBC.
+
+## General Concepts
+
+**IBC Enabled** - when instantiating a contract, we detect if it supports IBC messages
+  (this is an optional extension). If so, it is considered "IBC Enabled".
+  This info should be public on all contracts. (For mock, we assume all
+  contracts are IBC enabled)
+  
+Also, please read the [IBC Docs](https://docs.cosmos.network/master/ibc/overview.html)
+for detailed descriptions of the terms *Port*, *Client*, *Connection*,
+and *Channel*
+  
+## Workflow
+
+Establishing *Clients* and *Connections* is out of the scope of this
+module and must be created by the same means as for `ibc-transfer`
+(via the cli or otherwise). `x/wasm` will bind one or more *Ports* to
+handle messages for the contracts. How *Channels* are established
+is an open question for discussion.
+
+The following actions are only available for `IBCEnabled` contracts:
+
+* Dispatch `IBCSendMsg` - this sends an IBC packet over an established
+  channel. This is returned from `handle` just like any other `CosmosMsg`
+  (For mocks, we will trigger this externally, later only
+  valid contract addresses should be able to do so).
+* Handle `IBCRecvPacket` - when another chain sends a packet destined to
+  an IBCEnabled contract, this will be routed to the proper contract
+  and call an exposed function for this purpose.
+* Handle `IBCAckPacket` - The original sender of `IBCSendMsg` will
+  get this callback eventually if the message was successfully
+  processed on the other chain
+* Handle `IBCErrorPacket` - The original sender of `IBCSendMsg` will 
+  get this callback eventually if the message failed to be
+  processed on the other chain (for any reason)
+
```

Comment:
We need to let contracts handle `timeouts` where the packet was not delivered in time for any reason so that they can revert their operation